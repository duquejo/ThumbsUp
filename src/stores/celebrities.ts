import { computed, reactive, type ComputedRef } from 'vue';
import { defineStore } from 'pinia';
import { data } from '@/data.json';
import type { ICelebrity } from '@/interfaces/celebrities';

export type VoteTypes = 'positive' | 'negative';

export const useCelebritiesStore = defineStore('celebrities', () => {
  const celebrities = reactive<ICelebrity[]>(data);

  const getRandomCelebrity = computed<ICelebrity>(() => {
    const rand = Math.floor(Math.random() * celebrities.length);
    return celebrities[rand];
  });

  return {
    // State
    celebrities, 
    
    // Actions
    celebrityVote: (id: number, vote: VoteTypes) => {
      celebrities.map((celeb: ICelebrity) => {
        if (celeb.id === id) {
          return {
            ...celeb,
            votes: {
              positive: vote === 'positive' ? celeb.votes.positive++ : celeb.votes.positive,
              negative: vote === 'negative' ? celeb.votes.negative++ : celeb.votes.negative,
            },
          };
        }
        return celeb;
      });
    },

    getCelebrityStatusById: (id: number) => {
      const celebrity = celebrities.find((celeb: ICelebrity) => celeb.id === id);
      if (!celebrity) throw new Error('Celebrity not found');
      return celebrity.votes.positive >= celebrity.votes.negative ? 'positive' : 'negative';
    },
    
    // Getters
    getRandomCelebrity
  };
});
