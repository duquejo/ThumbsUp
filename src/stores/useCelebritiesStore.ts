import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { ICelebrity } from '@/interfaces/Celebrities';
import useLocalStorage from '@/composables/useLocalStorage';
import { postCelebrityVote } from '@/api/celebritiesApi';

export type VoteTypes = 'positive' | 'negative';

export const useCelebritiesStore = defineStore('celebrities', () => {
  const { saveItem } = useLocalStorage('celebrities');

  const celebrities = ref<ICelebrity[]>([]);

  return {
    // State
    celebrities,

    // Actions
    celebrityVote: (id: number, vote: VoteTypes) => {
      celebrities.value.map((celeb: ICelebrity) => {
        if (celeb.id === id) {
          return {
            ...celeb,
            votes: {
              positive: vote === 'positive' ? celeb.votes.positive++ : celeb.votes.positive,
              negative: vote === 'negative' ? celeb.votes.negative++ : celeb.votes.negative,
            },
          };
        }
        return celeb;
      });
      if( import.meta.env.VITE_VOTES_URL ) {
        postCelebrityVote(id, vote);
      }
      saveItem(celebrities.value);
    },
    setCelebritiesState: (newCelebs: ICelebrity[]) => {
      celebrities.value = newCelebs;
      saveItem(newCelebs);
    },
  };
});
