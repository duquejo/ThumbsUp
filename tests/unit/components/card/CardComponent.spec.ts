import { VueWrapper, shallowMount } from '@vue/test-utils';
import CardComponentVue from '@/components/card/CardComponent.vue';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import { useCelebritiesStore } from '@/stores/celebrities';
import { data } from '@/data.json'

vi.mock('@/stores/celebrities', () => ({
  useCelebritiesStore: vi.fn(() => ({
    getCelebrityStatusById: (id: number) => {
      return {
        value: 'positive',
      };
    },
  }))
}));

describe('+ CardComponent unit tests', () => {

  const properties = {
    celebrity: data[0]
  };

  let wrapper: VueWrapper;

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = shallowMount(CardComponentVue, {
      props: {  ...properties },
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn(),
        })],
      },
    });
  });

  it('should render properly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should set by default grid mode - No class added', () => {
    const container = wrapper.find('.layered');
    expect(container.exists()).toBeFalsy();
  });

  it('should set by default grid mode - Layered class added', () => {
    const layeredWrapper = shallowMount(CardComponentVue, {
      props: {  ...properties, isLayered: true },
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn(),
        })],
      },
    });
    expect(layeredWrapper.html()).toMatchSnapshot();

    const container = layeredWrapper.find('.layered');

    expect(container.exists()).toBeTruthy();
  });
});