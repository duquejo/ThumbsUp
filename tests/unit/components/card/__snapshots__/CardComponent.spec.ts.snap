// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ CardComponent unit tests > should render properly 1`] = `
"<li data-v-d766fe8c="" class="card card__background" style="background-size: cover;">
  <!-- General card container -->
  <div data-v-d766fe8c="" class="card__container"><span data-v-d766fe8c="" class="card__icon"><!--v-if--></span><!-- Central card container -->
    <div data-v-d766fe8c="" class="card__content">
      <!-- Main info -->
      <div data-v-d766fe8c="" class="card__content-top"><span data-v-d766fe8c="" class="card__title">Pope Francis</span>
        <p data-v-d766fe8c="" class="card__description truncate-overflow">He’s talking tough on clergy sexual abuse, or is he just another pervert protector? or a true pedophile punishing pontiff?</p>
      </div><!-- Actions -->
      <div data-v-d766fe8c="" class="card__content-bottom">
        <!-- Vote action container -->
        <thumb-buttons-component-stub data-v-d766fe8c="" lastupdated="2020-03-10T23:08:57.892Z" category="religion" id="0"></thumb-buttons-component-stub>
      </div>
    </div>
    <div data-v-d766fe8c="" class="card__gap"></div>
  </div><!-- Thumbs gauge percentage -->
  <thumb-gauge-component-stub data-v-d766fe8c="" votes="[object Object]"></thumb-gauge-component-stub>
</li>"
`;

exports[`+ CardComponent unit tests > should set by default grid mode - Layered class added 1`] = `
"<li data-v-d766fe8c="" class="card card__background layered" style="background-size: cover;">
  <!-- General card container -->
  <div data-v-d766fe8c="" class="card__container"><span data-v-d766fe8c="" class="card__icon"><!--v-if--></span><!-- Central card container -->
    <div data-v-d766fe8c="" class="card__content">
      <!-- Main info -->
      <div data-v-d766fe8c="" class="card__content-top"><span data-v-d766fe8c="" class="card__title">Pope Francis</span>
        <p data-v-d766fe8c="" class="card__description truncate-overflow">He’s talking tough on clergy sexual abuse, or is he just another pervert protector? or a true pedophile punishing pontiff?</p>
      </div><!-- Actions -->
      <div data-v-d766fe8c="" class="card__content-bottom">
        <!-- Vote action container -->
        <thumb-buttons-component-stub data-v-d766fe8c="" lastupdated="2020-03-10T23:08:57.892Z" category="religion" id="0"></thumb-buttons-component-stub>
      </div>
    </div>
    <div data-v-d766fe8c="" class="card__gap"></div>
  </div><!-- Thumbs gauge percentage -->
  <thumb-gauge-component-stub data-v-d766fe8c="" votes="[object Object]"></thumb-gauge-component-stub>
</li>"
`;
